### PR TITLE
Correct egress IP assignment for egress node IPs

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
When writing e2e tests for Egress IP (upcoming PR soon), I realized a mistake made on my part in the assignment procedure, namely: when specifying an egress IP which is a cluster node X's IP address, we might assign it to another node Y in the assignment procedure. This patch thus fixes this by implementing a check and making sure that if the egress IP specified is a cluster node's, that it is also assigned to exactly that node. If that node does not have the egress label, then we do not perform an assignment and notify the user by triggering an event.  

Some of the tests in `egressip_test.go` were explicitly wrong and testing and verifying the complete opposite to what they should have....I obviously should have paid more attention to that and not let it slip by. 

/assign @danwinship 

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by danwinship** [comment]
> when specifying an egress IP which is a cluster node X's IP address, we might assign it to another node Y in the assignment procedure. This patch thus fixes this by implementing a check and making sure that if the egress IP specified is a cluster node's, that it is also assigned to exactly that node.

Hm... in openshift-sdn it's not valid to use a node's actual IP as an egress IP. Because the node IP is going to be used for random traffic from other pods on that node, so it's not useful as an identifiable egress IP

**Comment by alexanderConstantinescu** [comment]
> Hm... in openshift-sdn it's not valid to use a node's actual IP as an egress IP. Because the node IP is going to be used for random traffic from other pods on that node, so it's not useful as an identifiable egress IP

Dohh (insert pic of Homer Simpson) 

Do we still want to insure ourselves for such a scenario (i.e, user specifies an egress IP which happens to be a cluster node IP), or does that fall under the "don't do that" category?

**Comment by danwinship** [comment]
We should check for it and consider it unsatisfiable/an error, like whatever we do if they request an egress IP that's not in any of the available CIDR ranges.

**Comment by alexanderConstantinescu** [comment]
> We should check for it and consider it unsatisfiable/an error, like whatever we do if they request an egress IP that's not in any of the available CIDR ranges.

FYI @danwinship : I've update the PR to do this.

**Comment by alexanderConstantinescu** [comment]
@danwinship : could you please take a look at this again?

File: go-controller/pkg/ovn/egressip.go [hunk 1]
```diff
@@ -169,6 +173,16 @@ func (oc *Controller) syncEgressIPs(eIPs []interface{}) {
 	}
 }
 
+func (oc *Controller) isAnyClusterNodeIP(ip net.IP) *eNode {
+	for _, eNode := range oc.eIPAllocator {
+		if (utilnet.IsIPv6(ip) && eNode.v6IP != nil && eNode.v6IP.Equal(ip)) ||
+			(!utilnet.IsIPv6(ip) && eNode.v4IP != nil && eNode.v4IP.Equal(ip)) {
```
Comment (danwinship) [inline go-controller/pkg/ovn/egressip.go:179]: fwiw just `if ip.Equal(eNode.v6IP) || ip.Equal(eNode.v4IP)` would work.

File: go-controller/pkg/ovn/egressip.go [hunk 2]
```diff
@@ -279,24 +293,37 @@ func (oc *Controller) assignEgressIPs(eIP *egressipv1.EgressIP) error {
...
-				klog.V(5).Infof("Node: %s is already in use by another egress IP for this EgressIP: %s, trying another node", eNodes[i].name, eIP.Name)
+			if assignableNodes[i].tainted {
+				klog.V(5).Infof("Node: %s is already in use by another egress IP for this EgressIP: %s, trying another node", assignableNodes[i].name, eIP.Name)
 				continue
 			}
-			if (utilnet.IsIPv6(eIPC) && eNodes[i].v6Subnet != nil && eNodes[i].v6Subnet.Contains(eIPC)) ||
-				(!utilnet.IsIPv6(eIPC) && eNodes[i].v4Subnet != nil && eNodes[i].v4Subnet.Contains(eIPC)) {
-				eNodes[i].tainted, oc.eIPAllocator[eNodes[i].name].allocations[eIPC.String()] = true, true
+			if (utilnet.IsIPv6(eIPC) && assignableNodes[i].v6Subnet != nil && assignableNodes[i].v6Subnet.Contains(eIPC)) ||
+				(!utilnet.IsIPv6(eIPC) && assignableNodes[i].v4Subnet != nil && assignableNodes[i].v4Subnet.Contains(eIPC)) {
```
Comment (danwinship) [inline go-controller/pkg/ovn/egressip.go:319]: similarly, could do just

	if (assignableNodes[i].v6Subnet != nil && assignableNodes[i].v6Subnet.Contains(eIPC)) ||
		(assignableNodes[i].v4Subnet != nil && assignableNodes[i].v4Subnet.Contains(eIPC)) {

You're allowed to ask if an IPv6 subnet contains an IPv4 IP or vice versa. (The answer is just guaranteed to be "no".)

**Review by danwinship — COMMENTED** [review]
Some comments but I'm guess they probably apply to a lot of the existing code as well, and could maybe be addressed in a general cleanup later

**Pull request merged**